### PR TITLE
[IMP] base_import, web: add drag-and-drop support for import record view

### PR DESCRIPTION
**Pull request by julien-banken**
To simplify the process of importing records into Odoo, we will turn the import view into a dropzone. This will allow users to import records by dragging a file from their computer directly into the import view. This enhancement simplifies the import process, making it more convenient and user-friendly.

Even after a file is imported, the drag-and-drop feature will remain active, allowing users to drag and drop a new file to load new records.

task-3956293

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/base_import/static/src/import_action/import_action.js
Comment (awa-odoo) [inline addons/base_import/static/src/import_action/import_action.js:62]: You seem to also accept .xls

File: addons/base_import/static/src/import_action/import_action.js
Comment (awa-odoo) [inline addons/base_import/static/src/import_action/import_action.js:55]: Don't really understand how that turns only the no-content-helper into a drop zone, seems to use the whole component.

This allows re-dropping a file even when you're importing, which is cool I guess, but not really implied in commit message.

File: addons/base_import/static/tests/import_action_tests.js
Comment (awa-odoo) [inline addons/base_import/static/tests/import_action_tests.js:571]: "o-mail-Dropzone"? Seems like a mail-specific class, not sure what it has to do with anything here.

File: addons/web/static/src/core/file_input/file_input.js
Comment (awa-odoo) [inline addons/web/static/src/core/file_input/file_input.js:118]: Shouldn't this be inside "core/utils/files" along with "checkFileSize" ?

**Review by awa-odoo — COMMENTED** [review]
Hello @julien-banken 

Neat little feature, left a few comments.
Once finished, we'll try to get a testing round from Luc (wording etc), but then we'll be able to merge right away normally, no need for a second review phase.

Thanks fro your work & cheers!

File: addons/base_import/static/src/import_action/import_action.js
Comment (julien-banken) [reply addons/base_import/static/src/import_action/import_action.js:55]: The no-content helper element is too small and I think it's not where we should set the dropzone. Ideally, I would like to set the dropzone on the view gray zone but it involves more changes because the view gray zone is defined inside the generic `Layout` component.

Maybe, I can define a variant of the `Layout` component, plug the hook I want and pass a callback function `onDrop` to upload the file but I feel like it was a bit overkill. So, I just put it there on the root node of the view.

**Do you prefer having the dropzone on the view gray zone?** I can change that if you prefer :smiley:

If it's okay, I will update the commit message.

File: addons/base_import/static/src/import_action/import_action.js
Comment (julien-banken) [reply addons/base_import/static/src/import_action/import_action.js:56]: If the user drops multiple files, I prefer picking the first valid file instead of showing an error. It makes the system more flexible. I can change that as well if you prefer an error :smiley:

File: addons/base_import/static/src/import_action/import_action.js
Comment (julien-banken) [reply addons/base_import/static/src/import_action/import_action.js:57]: As a wise man would say: "boarf :boar:". Here, I call one function instead of chaining two functions. The function is standard and self-explanatory: It returns true if the string `field.name` ends with `.csv`.

File: addons/base_import/static/tests/import_action_tests.js
Comment (julien-banken) [reply addons/base_import/static/tests/import_action_tests.js:571]: It's because the `useDropzone` is defined in the `mail` module and the discuss team uses a nomenclature "<module_name>-<component_name>" for their class names. Maybe this has to be moved to `web`?

**Actually, I should maybe do the move now because `base_import` depends on `web` only and `web` depends on `base` only and `base` has no dependency. So, it's possible that the import will be undefined as we may not have `mail` installed.**

File: addons/base_import/static/src/import_action/import_action.js
Comment (awa-odoo) [reply addons/base_import/static/src/import_action/import_action.js:56]: I think it's fine to go for this, kind of cool that you can re-upload something multiple times by drag & dropping IMO.

So just changing the commit message should be sufficient 🙂

File: addons/base_import/static/src/import_action/import_action.js
Comment (awa-odoo) [reply addons/base_import/static/src/import_action/import_action.js:57]: Meh, I prefer a clear error message, people may drag & drop 2 import files and think the system is totally fine with that, which is not the case.

File: addons/web/static/src/core/utils/files.js
Comment (awa-odoo) [inline addons/web/static/src/core/utils/files.js:36]: Can I use "useService" inside hooks? I thought not as "checkFileSize" is requesting the "notificationService" as a method argument, but maybe it works 🤷‍♂️

File: addons/mail/static/tests/composer/composer.test.js
Comment (alexkuhn) [inline addons/mail/static/tests/composer/composer.test.js:586]: Components in `web` can omit the moduleName, so it just becomes `o-Dropzone`

File: addons/web/static/src/core/utils/files.js
Comment (alexkuhn) [reply addons/web/static/src/core/utils/files.js:36]: Yes `useSerice()` inside hooks is good. As long as there's no dependency cycle and the hook is called in `Component.setup()`, you can have hooks call other hooks

File: addons/base_import/static/tests/import_action_tests.js
Comment (julien-banken) [inline addons/base_import/static/tests/import_action_tests.js]: I will rather use the `dragenterFiles` and `dropFiles` helper functions to do that.

File: addons/base_import/static/tests/import_action_tests.js
Comment (awa-odoo) [inline addons/base_import/static/tests/import_action_tests.js:14]: A bit strange to use legacy helper but I'm guessing those will get converted at some point.

**Comment by awa-odoo** [comment]
@odoo/rd-framework-js FYI, slight change in the FileInput component, don't think it deserves your attention.

**Review by awa-odoo — APPROVED** [review]
LGTM, let's go to testing!

File: addons/base_import/static/tests/import_action_tests.js
Comment (aab-odoo) [reply addons/base_import/static/tests/import_action_tests.js:14]: This suite is still legacy. @julien-banken this could be the time to convert it to hoot (not in this PR, but after)

Comment (julien-banken) [reply addons/base_import/static/tests/import_action_tests.js:14]: Ok, we can do that in another PR. As there are no hoot tests for the `import_action` module yet, I've simply added a legacy test to speed things up. I suggest to migrate all of the tests at once when we have more time.

**Comment by awa-odoo** [comment]
@julien-banken Hello

Small file conflict on this one apparently, when you have time would be nice to rebase so we can close 🙂 

Thanks & cheers!

**Comment by awa-odoo** [comment]
@robodoo r+

**Pull request closed without merging**
